File: client/src/main/java/io/littlehorse/jlib/worker/LHTaskWorker.java
package io.littlehorse.jlib.worker;

import io.littlehorse.jlib.client.LHClient;
import io.littlehorse.jlib.common.LHLibUtil;
import io.littlehorse.jlib.common.config.LHWorkerConfig;
import io.littlehorse.jlib.common.exception.LHApiError;
import io.littlehorse.jlib.common.exception.TaskSchemaMismatchError;
import io.littlehorse.jlib.common.proto.LHResponseCodePb;
import io.littlehorse.jlib.common.proto.TaskDefPb;
import io.littlehorse.jlib.common.proto.VariableTypePb;
import io.littlehorse.jlib.wfsdk.internal.taskdefutil.LHTaskSignature;
import io.littlehorse.jlib.wfsdk.internal.taskdefutil.TaskDefBuilder;
import io.littlehorse.jlib.worker.internal.LHServerConnectionManager;
import io.littlehorse.jlib.worker.internal.util.VariableMapping;
import java.io.Closeable;
import java.io.IOException;
import java.lang.reflect.Method;
import java.lang.reflect.Parameter;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * The LHTaskWorker talks to the LH Servers and executes a specified Task Method every
 * time a Task is scheduled.
 */
public class LHTaskWorker implements Closeable {

    private Logger log = LoggerFactory.getLogger(LHTaskWorker.class);

    public static HashMap<Class<?>, VariableTypePb> javaTypeToLHType = new HashMap<>() {
        {
            put(Integer.class, VariableTypePb.INT);
            put(Long.class, VariableTypePb.INT);
            put(Boolean.class, VariableTypePb.BOOL);
            put(Double.class, VariableTypePb.DOUBLE);
            put(byte[].class, VariableTypePb.BYTES);
            put(String.class, VariableTypePb.STR);
        }
    };

    private Object executable;
    private LHWorkerConfig config;
    private TaskDefPb taskDef;
    private Method taskMethod;
    private List<VariableMapping> mappings;
    private LHServerConnectionManager manager;
    private LHClient lhClient;
    private String taskDefName;

    /**
     * Creates an LHTaskWorker given an Object that has an annotated LHTaskMethod,
     * and a configuration Properties object.
     * @param executable is any Object which has exactly one method annotated with
     * '@LHTaskMethod'. That method will be used to execute the tasks.
     * @param taskDefName is the name of the `TaskDef` to execute.
     * @param config is a valid LHWorkerConfig.
     */
    public LHTaskWorker(
        Object executable,
        String taskDefName,
        LHWorkerConfig config
    ) {
        this.config = config;
        this.executable = executable;
        this.mappings = new ArrayList<>();
        this.lhClient = new LHClient(config);
        this.taskDefName = taskDefName;
    }

    private void createManager() throws LHApiError {
        try {
            validateTaskDefAndExecutable();

            this.manager =
                new LHServerConnectionManager(
                    taskMethod,
                    taskDef,
                    config,
                    mappings,
                    executable
                );
        } catch (TaskSchemaMismatchError exn) {
            throw new LHApiError(
                exn,
                "Provided java method does not match registered task!",
                LHResponseCodePb.BAD_REQUEST_ERROR
            );
        } catch (IOException exn) {
            throw new LHApiError(exn, "Couldn't create connection to LH");
        }
    }

    private void checkTaskDefExists() throws LHApiError {
        this.taskDef = lhClient.getTaskDef(taskDefName);

        if (this.taskDef == null) {
            throw new LHApiError(
                "Couldn't find TaskDef: " + taskDefName,
                LHResponseCodePb.NOT_FOUND_ERROR
            );
        }
    }

    /**
     * Deploys the  TaskDef object to the LH Server.
     * This is a convenience method, generally not recommended for
     * production (in production you should manually use the PutTaskDef).
     * @param swallowAlreadyExists if true, then ignore ALREADY_EXISTS_ERROR when
     * registering the TaskDef.
     * @throws LHApiError if the call fails.
     */
    public void registerTaskDef(boolean swallowAlreadyExists) throws LHApiError {
        try {
            TaskDefBuilder tdb = new TaskDefBuilder(executable, taskDefName);
            log.info(
                "Creating TaskDef:\n {}",
                LHLibUtil.protoToJson(
                    lhClient.putTaskDef(tdb.toPutTaskDefPb(), swallowAlreadyExists)
                )
            );
        } catch (TaskSchemaMismatchError exn) {
            log.error("Error registering task", exn);
            throw new LHApiError(
                exn,
                exn.getMessage(),
                LHResponseCodePb.VALIDATION_ERROR
            );
        }
    }

    private void validateTaskDefAndExecutable() throws TaskSchemaMismatchError {
        LHTaskSignature signature = new LHTaskSignature(
            taskDef.getName(),
            executable
        );
        taskMethod = signature.getTaskMethod();

        int numTaskMethodParams = taskMethod.getParameterCount();
        int numTaskDefParams = taskDef.getInputVarsCount();

        boolean wrongNumParams = false;
        if (signature.getHasWorkerContextAtEnd()) {
            if (numTaskMethodParams - 1 != numTaskDefParams) {
                wrongNumParams = true;
            }
        } else if (numTaskDefParams != numTaskMethodParams) {
            wrongNumParams = true;
        }

        if (wrongNumParams) {
            throw new TaskSchemaMismatchError(
                "Number of task method params doesn't match number of taskdef params!"
            );
        }

        for (int i = 0; i < numTaskDefParams; i++) {
            Parameter param = taskMethod.getParameters()[i];
            String javaParamName = param.getName();
            Class<?> paramClass = param.getType();

            if (paramClass.equals(WorkerContext.class)) {
                throw new TaskSchemaMismatchError(
                    "Can only have WorkerContext after all required taskDef params."
                );
            }

            // This line throws a TaskSchemaMismatchError if the param can't
            // be provided properly.
            VariableMapping mapping = new VariableMapping(
                taskDef,
                i,
                paramClass,
                javaParamName
            );
            mappings.add(mapping);
        }

        if (signature.getHasWorkerContextAtEnd()) {
            mappings.add(
                new VariableMapping(
                    taskDef,
                    numTaskMethodParams - 1,
                    WorkerContext.class,
                    null
                )
            );
        }
    }

    /**
     * Starts polling for and executing tasks.
     * @throws LHApiError if the schema from the TaskDef configured in
     * the configProps is incompatible with the method signature from the provided
     * executable Java object, or if the Worker cannot connect to the LH Server.
     */
    public void start() throws LHApiError {
        checkTaskDefExists();
        createManager();
        manager.start();
    }

    /**
     * Cleanly shuts down the Task Worker.
     */
    public void close() {
        manager.close();
    }
}

File: sdk-dotnet/LittleHorseSDK/LittleHorse.Worker/LHTaskWorker.cs
using Grpc.Core;
using LittleHorse.Common.Configuration;
using LittleHorse.Common.Exceptions;
using LittleHorse.Worker.Internal;
using LittleHorse.Worker.Internal.Helpers;
using LittleHorse.Common.Proto;
using Microsoft.Extensions.Logging;
using System.Reflection;

namespace LittleHorse.Worker
{
    /// <summary>
    /// The LHTaskWorker talks to the LH Servers and executes a specified Task Method every time a Task
    /// is scheduled.
    /// </summary>
    /// <typeparam name="T">
    /// The type of the executable object.
    /// </typeparam>
    public class LHTaskWorker<T>
    {
        private ILHWorkerConfig _config;
        private ILogger<LHTaskWorker<T>>? _logger;
        private T _executable;
        private TaskDef? _taskDef;
        private MethodInfo? _taskMethod;
        private List<VariableMapping> _mappings;
        private LHTaskSignature<T>? _taskSignature;
        private LHServerConnectionManager<T>? _manager;
        private string _taskDefName;
        private LHPublicApi.LHPublicApiClient _grpcClient;

        public string TaskDefName { get => _taskDefName; }

        public LHTaskWorker(T executable, string taskDefName, ILHWorkerConfig config, ILogger<LHTaskWorker<T>>? logger = null)
        {
            _config = config;
            _logger = logger;
            _executable = executable;
            _mappings = new List<VariableMapping>();
            _taskDefName = taskDefName;
            _grpcClient = _config.GetGrcpClientInstance();
        }

        /// <summary>
        /// Starts polling for and executing tasks.
        /// </summary>
        /// <exception cref="LHMisconfigurationException">
        /// if the schema from the TaskDef configured in the configProps is
        /// incompatible with the method signature from the provided executable Java object, or if
        /// the Worker cannot connect to the LH Server.
        /// </exception>
        public void Start()
        {
            if (!TaskDefExists())
            {
                throw new LHMisconfigurationException($"Couldn't find TaskDef: {_taskDefName}");
            }

            _taskSignature = new LHTaskSignature<T>(_taskDefName, _executable);
            _taskMethod = _taskSignature.TaskMethod;

            ValidateTaskMethodParameters(_taskMethod, _taskSignature);
            _mappings = CreateVariableMappings(_taskMethod, _taskSignature);

            _manager = new LHServerConnectionManager<T>(_config, _taskMethod, GetTaskDef(), _mappings, _executable, _logger);

            _manager.Start();
        }

        /// <summary>
        /// Checks if the TaskDef exists
        /// </summary>
        /// <returns>
        /// true if the task is registered or false otherwise
        /// </returns>
        /// <exception cref="RpcException"> Throws when call fails.
        /// </exception>
        public bool TaskDefExists()
        {
            try
            {
                var taskDefId = new TaskDefId()
                {
                    Name = _taskDefName,
                };
                _grpcClient.GetTaskDef(taskDefId);

                return true;
            }
            catch (RpcException ex)
            {
                if (ex.StatusCode == StatusCode.NotFound)
                {
                    return false;
                }

                throw;
            }
        }

        /// <summary>
        /// Cleanly shuts down the Task Worker.
        /// </summary>
        public void Close()
        {
            _manager?.Dispose();
        }

        /// <summary>
        /// Deploys the TaskDef object to the LH Server. This is a convenience method, generally not
        /// recommended for production(in production you should manually use the PutTaskDef).
        /// </summary>
        public void RegisterTaskDef()
        {
            RegisterTaskDef(false);
        }

        /// <summary>
        /// Deploys the TaskDef object to the LH Server. This is a convenience method, generally not
        /// recommended for production(in production you should manually use the PutTaskDef).
        /// </summary>
        /// <param name="swallowAlreadyExists">
        /// If true, then ignore grpc ALREADY_EXISTS error when registering the TaskDef.
        /// </param>
        public void RegisterTaskDef(bool swallowAlreadyExists)
        {
            _logger?.LogInformation($"Creating TaskDef: {_taskDefName}");

            try
            {
                var signature = new LHTaskSignature<T>(_taskDefName, _executable);

                var request = new PutTaskDefRequest()
                {
                    Name = _taskDefName
                };

                for (int i = 0; i < signature.VarNames.Count; i++)
                {
                    request.InputVars.Add(new VariableDef()
                    {
                        Name = signature.VarNames[i],
                        Type = signature.ParamTypes[i]
                    });
                }

                var response = _grpcClient.PutTaskDef(request);
                _logger?.LogInformation($"Created TaskDef:\n{LHMappingHelper.MapProtoToJson(response)}");
            }
            catch (RpcException ex)
            {
                if (swallowAlreadyExists && ex.StatusCode == StatusCode.AlreadyExists)
                {
                    _logger?.LogInformation($"TaskDef {_taskDefName} already exists!");
                }
                else
                {
                    throw;
                }
            }
        }

        private TaskDef GetTaskDef()
        {
            if (_taskDef is null)
            {
                _taskDef = _config.GetTaskDef(_taskDefName);
            }

            return _taskDef;
        }

        private void ValidateTaskMethodParameters(MethodInfo taskMethod, LHTaskSignature<T> taskSignature)
        {

            if (taskSignature.HasWorkerContextAtEnd)
            {
                if (taskSignature.TaskMethod.GetParameters().Length - 1 != GetTaskDef().InputVars.Count)
                {
                    throw new LHTaskSchemaMismatchException("Number of task method params doesn't match number of taskdef params!");
                }
            }
            else
            {
                if (taskMethod.GetParameters().Length != GetTaskDef().InputVars.Count)
                {
                    throw new LHTaskSchemaMismatchException("Number of task method params doesn't match number of taskdef params!");
                }
            }
        }

        private List<VariableMapping> CreateVariableMappings(MethodInfo taskMethod, LHTaskSignature<T> taskSignature)
        {
            var mappings = new List<VariableMapping>();

            var taskParams = taskMethod.GetParameters();

            for (int index = 0; index < GetTaskDef().InputVars.Count; index++)
            {
                var taskParam = taskParams[index];

                if (taskParam.ParameterType.IsAssignableFrom(typeof(LHWorkerContext)))
                {
                    throw new LHTaskSchemaMismatchException("Can only have WorkerContext after all required taskDef params.");
                }

                mappings.Add(CreateVariableMapping(GetTaskDef(), index, taskParam.ParameterType, taskParam.Name));
            }

            if (taskSignature.HasWorkerContextAtEnd)
            {
                mappings.Add(CreateVariableMapping(GetTaskDef(), taskParams.Count() - 1, typeof(LHWorkerContext), null));
            }

            return mappings;
        }

        private VariableMapping CreateVariableMapping(TaskDef taskDef, int index, Type type, string? paramName)
        {
            return new VariableMapping(taskDef, index, type, paramName);
        }

    }
}
